File: src/net/eduard/api/server/minigame/Minigame.java
package net.eduard.api.server.minigame;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import net.eduard.api.lib.storage.StorageAttributes;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import net.eduard.api.lib.Mine;
import net.eduard.api.lib.game.DisplayBoard;
import net.eduard.api.lib.game.Kit;
import net.eduard.api.lib.game.MinigameChest;
import net.eduard.api.lib.manager.TimeManager;
import net.eduard.api.lib.modules.BukkitBungeeAPI;

/**
 * Representa um Jogo <br>
 * MinigameSetup 1.0
 * 
 * @version 2.0
 * @since EduardAPI 2.0
 * @author Eduard
 *
 */
@StorageAttributes(indentificate = true)
public class Minigame extends TimeManager {

	private String name = "Minigame";
	private String messagePrefix = "[Minigame] ";
	private boolean enabled = true;
	private boolean bungeecord = true;
	private String bungeeLobby = "Lobby";
	private Location lobby;
	private int maxPlayersPerLobby = 20;
	private int timeIntoStart = 60;
	private int timeIntoRestart = 20;
	private int timeIntoGameOver = 15 * 60;
	private int timeIntoPlay = 2 * 60;

	private int timeOnStartTimer = 0;
	private int timeOnRestartTimer = 40;
	private int timeOnForceTimer = 10;
	private int timeOnStartingToBroadcast = 15;
	private int timeOnEquipingToBroadcast = 1;
	private transient MinigameMap setting;
	private transient List<MinigamePlayer> players = new ArrayList<>();
	private DisplayBoard scoreboardStarting;
	private DisplayBoard scoreboardLobby;
	private DisplayBoard scoreboardPlaying;
	private MinigameChest chests = new MinigameChest();
	private MinigameChest chestsFeast = new MinigameChest();
	private MinigameChest chestMiniFeast = new MinigameChest();
	private List<Kit> kits = new ArrayList<>();
	private List<MinigameLobby> lobbies = new ArrayList<>();
	private List<MinigameMap> maps = new ArrayList<>();
	private List<MinigameRoom> rooms = new ArrayList<>();

	/**
	 * Conecta todos jogadores no servidor Lobby
	 */
	public void connectAllPlayersToLobby() {
		for (MinigamePlayer player : players) {
			BukkitBungeeAPI.connectToServer(player.getPlayer(), bungeeLobby);
		}

	}

	/**
	 * Teleporta todos os jogadores para o Local do Lobby
	 */
	public void teleportAllPlayersToLobby() {
		getPlayers().forEach(p -> p.getPlayer().teleport(getLobby()));
//		for (MinigamePlayer player : getPlayers()) {
//			Player p = player.getPlayer();
//			p.teleport(getLobby());
//		}
	}

	public Minigame() {
	}

	public int getTimeIntoPlay() {
		return timeIntoPlay;
	}

	public void setTimeIntoPlay(int timeIntoPlay) {
		this.timeIntoPlay = timeIntoPlay;
	}

	public Minigame(String name) {
		setName(name);
		setMessagePrefix("§8[§b" + name + "§8] ");
		lobbies.add(new MinigameLobby());
	}

	/**
	 * Cria um Mapa
	 * 
	 * @param nome Nome
	 * @return Mapa Novo
	 */
	public MinigameMap createMap(String nome) {
		MinigameMap map = new MinigameMap(nome);
		getMaps().add(map);
		return map;

	}

	/**
	 * Timer do Minigame define oque acontece a cada segundo que se passa do
	 * Minigame em cada Sala
	 * 
	 * @param room Sala
	 */
	public void event(MinigameRoom room) {
	}

	/**
	 * Pega o mapa existente pelo seu nome
	 * 
	 * @param name Nome
	 * @return Mapa
	 */
	public MinigameMap getMap(String name) {
		for (MinigameMap map : getMaps()) {
			if (map.getName().equalsIgnoreCase(name)) {
				return map;
			}
		}
		return null;
	}

	/**
	 * Remove o mapa da lista de mapas existentes
	 * 
	 * @param map
	 */
	public void removeMap(MinigameMap map) {
		getMaps().remove(map);
	}

	/**
	 * Verifica se existe este Map com este Nome
	 * 
	 * @param name Nome
	 * @return
	 */
	public boolean hasMap(String name) {
		for (MinigameMap map : getMaps()) {
			if (map.getName().equalsIgnoreCase(name)) {
				return true;
			}
		}
		return false;
	}

	/**
	 * Cria um sala unica e um mapa unico Também usando o Nome do Minigame
	 * 
	 * @return Minigame criado com mapa já configurado
	 */
	public MinigameRoom uniqueGame() {
		return new MinigameRoom(this, new MinigameMap(this, getName()));
	}

	public Minigame(String name, Plugin plugin) {
		this(name);
		setPlugin(plugin);
	}

	/**
	 * Manda mensagem para todos jogadores participando do minigame
	 * 
	 * @param message Mensagem
	 */
	public void broadcast(String message) {
		for (Player player : getPlayersOnline()) {
			player.sendMessage(messagePrefix + Mine.getReplacers(message, player));
		}
	}

	public String getBungeeLobby() {
		return bungeeLobby;
	}

	/**
	 * Pega a primera sala existente do Minigame
	 * 
	 * @return Sala
	 */
	public MinigameRoom getGame() {
//		getRooms().get(0);
		return getRooms().iterator().next();
	}

	/**
	 * Pega a sala pelo seu ID
	 * 
	 * @param id ID
	 * @return Sala
	 */
	public MinigameRoom getRoom(int id) {
		for (MinigameRoom room : rooms) {
			if (room.getId() == id) {
				return room;
			}
		}
		return null;
	}

	/**
	 * Verifica se a sala com este ID existe
	 * 
	 * @param id ID
	 * @return Sala
	 */
	public boolean hasRoom(int id) {
		return getRoom(id) != null;
	}

	/**
	 * Cria uma Sala com este ID para o Mapa expecifico
	 * 
	 * @param map Mapa
	 * @param id  ID
	 * @return Nova Sala
	 */
	public MinigameRoom createRoom(MinigameMap map, int id) {
		MinigameRoom room = new MinigameRoom(this, map);
		room.setId(id);
		return room;

	}

	public List<MinigameRoom> getRooms() {
		return rooms;
	}

	public void setRooms(List<MinigameRoom> rooms) {
		this.rooms = rooms;
	}

	public List<MinigamePlayer> getPlayers() {
		return players;
	}

	public void setPlayers(List<MinigamePlayer> players) {
		this.players = players;
	}

	/**
	 * Pega a sala que o jogador esta jogando
	 * 
	 * @param player Jogador
	 * @return Sala do jogador
	 */
	public MinigameRoom getGame(Player player) {
		return getPlayer(player).getGame();
	}

	/**
	 * Pega a sala com o nome do seu mapa igual a este
	 * 
	 * @param name Nome
	 * @return
	 */
	public MinigameRoom getGame(String name) {
		for (MinigameRoom room : rooms) {
			if (room.getMap().getName().equalsIgnoreCase(name)) {
				return room;
			}
		}
		return null;
	}

	public Location getLobby() {
		return lobby;
	}

	public MinigameLobby getMainLobby() {
		if (lobbies.size() > 0)
			return lobbies.get(0);
		return newLobby(1);
	}

	public MinigameLobby newLobby(int id) {

		MinigameLobby lobby = new MinigameLobby();
		lobby.setId(id);
		lobbies.add(lobby);
		return lobby;

	}

	/**
	 * Pega o mapa referente a sala principal do Minigame
	 * 
	 * @return
	 */
	public MinigameMap getMap() {
		return getMap(getName());
	}

	public String getName() {
		return name;
	}

	/**
	 * Pega o MinigamePlayer referente ao jogador e se caso não exista cria um
	 * 
	 * @param player Jogador
	 * @return Instancia de MinigamePlayer (MP)
	 */
	public MinigamePlayer getPlayer(Player player) {
		MinigamePlayer miniplayer;
		for (MinigamePlayer p : players) {
			if (player.equals(p.getPlayer())) {
				return p;
			}
		}
		miniplayer = new MinigamePlayer(player);
		players.add(miniplayer);

		return miniplayer;
	}

	/**
	 * Pega os jogadores que estão jogando
	 * 
	 * @return Lista de Jogadores ({@link Player})
	 */
	public List<Player> getPlayersOnline() {
		List<Player> list = new ArrayList<>();
		for (MinigamePlayer player : players) {
			list.add(player.getPlayer());
		}

		// List<Player> listaDoPlayers =
		// players.stream().map(MinigamePlayer::getPlayer).collect(Collectors.toList());
		return list;
	}

	/**
	 * Verifica se existe o lobby
	 * 
	 * @return
	 */
	public boolean hasLobby() {
		return lobby != null;
	}

	/**
	 * Verifica se o jogador esta no modo Admin
	 * 
	 * @param player Jogador
	 * @return
	 */
	public boolean isAdmin(Player player) {
		return getPlayer(player).isState(MinigamePlayerState.ADMIN);

	}

	/**
	 * Verifica se é modo Bungeecord
	 * 
	 * @return
	 */
	public boolean isBungeecord() {
		return bungeecord;
	}

	/**
	 * Verifica se o minigame esta ativado
	 * 
	 * @return
	 */
	public boolean isEnabled() {
		return enabled;
	}

	/**
	 * Verifica se o Jogador esta no modo Normal (sem ser Admin ou Spectador)
	 * 
	 * @param player Jogador
	 * @return
	 */
	public boolean isPlayer(Player player) {
		return getPlayer(player).isState(MinigamePlayerState.NORMAL);

	}

	/**
	 * Verifica se o jogador esta no Minigame
	 * 
	 * @param player Jogador
	 * @return
	 */
	public boolean isPlaying(Player player) {
		return getPlayer(player).isPlaying();
	}

	/**
	 * Verifica se o Jogador esta no modo Spectador
	 * 
	 * @param player Jogador
	 * @return
	 */
	public boolean isSpectator(Player player) {
		return getPlayer(player).isState(MinigamePlayerState.SPECTATOR);

	}

	/**
	 * Verifica se o Estado da Sala principal é igual este estado
	 * 
	 * @param state Estado
	 * @return
	 */
	public boolean isState(MinigameState state) {
		return getGame().isState(state);
	}

	/**
	 * Entrar em uma Sala
	 * 
	 * @param game   Sala
	 * @param player Jogador
	 */
	public void joinPlayer(MinigameRoom game, Player player) {
		MinigamePlayer p = getPlayer(player);
		p.join(game);

	}

	/**
	 * Entrar em um Time
	 * 
	 * @param team   Time
	 * @param player Jogador
	 */
	public void joinPlayer(MinigameTeam team, Player player) {
		MinigamePlayer p = getPlayer(player);
		p.join(team);
	}

	/**
	 * Remover o jogador da sala e do time Atual dele
	 * 
	 * @param player Jogador
	 */
	public void leavePlayer(Player player) {
		MinigamePlayer p = getPlayer(player);
		if (p.isPlaying()) {
			p.getGame().leave(p);
		}

		if (p.hasTeam()) {
			p.getTeam().leave(p);
		}

	}

	/**
	 * Remove o jogador da Lista de jogadores {@link MinigamePlayer}
	 * 
	 * @param player Jogador
	 */
	public void remove(Player player) {
		players.remove(getPlayer(player));
	}

	/**
	 * Remove a Sala da lista de salas existentes
	 * 
	 * @param game Sala
	 */
	public void removeGame(MinigameRoom game) {
		this.rooms.remove(game);
	}

	/**
	 * Remove a sala pelo seu ID
	 * 
	 * @param id
	 */
	public void removeGame(int id) {
		MinigameRoom game = getRoom(id);
		this.rooms.remove(game);
	}

	public Object restore(Map<String, Object> map) {

		return null;
	}

	/**
	 * Metodo que é executado a cada segundo e executa o metodo de cada sala
	 * <code> events(room)</code>
	 */
	public void run() {
		if (!enabled)
			return;
		for (MinigameRoom room : rooms) {
			if (!room.isEnabled())
				continue;
			event(room);
		}
	}

	public void setBungeecord(boolean bungeecord) {
		this.bungeecord = bungeecord;
	}

	public void setBungeeLobby(String bungeeLobby) {
		this.bungeeLobby = bungeeLobby;
	}

	public void setEnabled(boolean enabled) {
		this.enabled = enabled;
	}

	public void setLobby(Location lobby) {
		this.lobby = lobby;
	}

	public void setName(String name) {
		this.name = name;
	}

	public MinigameMap getSetting() {
		return setting;
	}

	public boolean isSetting() {
		return setting != null;
	}

	public void setSetting(MinigameMap setting) {
		this.setting = setting;
	}

	public int getTimeIntoStart() {
		return timeIntoStart;
	}

	public void setTimeIntoStart(int timeIntoStart) {
		this.timeIntoStart = timeIntoStart;
	}

	public int getTimeIntoRestart() {
		return timeIntoRestart;
	}

	public void setTimeIntoRestart(int timeIntoRestart) {
		this.timeIntoRestart = timeIntoRestart;
	}

	public int getTimeIntoGameOver() {
		return timeIntoGameOver;
	}

	public void setTimeIntoGameOver(int timeIntoGameOver) {
		this.timeIntoGameOver = timeIntoGameOver;
	}

	public int getTimeWithoutPvP() {
		return timeIntoPlay;
	}

	public void setTimeWithoutPvP(int timeWithoutPvP) {
		this.timeIntoPlay = timeWithoutPvP;
	}

	public int getTimeOnStartTimer() {
		return timeOnStartTimer;
	}

	public void setTimeOnStartTimer(int timeOnStartTimer) {
		this.timeOnStartTimer = timeOnStartTimer;
	}

	public int getTimeOnRestartTimer() {
		return timeOnRestartTimer;
	}

	public void setTimeOnRestartTimer(int timeOnRestartTimer) {
		this.timeOnRestartTimer = timeOnRestartTimer;
	}

	public int getTimeOnForceTimer() {
		return timeOnForceTimer;
	}

	public void setTimeOnForceTimer(int timeOnForceTimer) {
		this.timeOnForceTimer = timeOnForceTimer;
	}

	public DisplayBoard getScoreboardLobby() {
		return scoreboardLobby;
	}

	public void setScoreboardLobby(DisplayBoard scoreboardLobby) {
		this.scoreboardLobby = scoreboardLobby;
	}

	public DisplayBoard getScoreboardPlaying() {
		return scoreboardPlaying;
	}

	public void setScoreboardPlaying(DisplayBoard scoreboardPlaying) {
		this.scoreboardPlaying = scoreboardPlaying;
	}

	public MinigameChest getChests() {
		return chests;
	}

	public void setChests(MinigameChest chests) {
		this.chests = chests;
	}

	public MinigameChest getChestsFeast() {
		return chestsFeast;
	}

	public void setChestsFeast(MinigameChest chestsFeast) {
		this.chestsFeast = chestsFeast;
	}

	public MinigameChest getChestMiniFeast() {
		return chestMiniFeast;
	}

	public void setChestMiniFeast(MinigameChest chestMiniFeast) {
		this.chestMiniFeast = chestMiniFeast;
	}

	public List<Kit> getKits() {
		return kits;
	}

	public void setKits(List<Kit> kits) {
		this.kits = kits;
	}

	public List<MinigameMap> getMaps() {
		return maps;
	}

	public void setMaps(List<MinigameMap> maps) {
		this.maps = maps;
	}

	public DisplayBoard getScoreboardStarting() {
		return scoreboardStarting;
	}

	public void setScoreboardStarting(DisplayBoard scoreboardStarting) {
		this.scoreboardStarting = scoreboardStarting;
	}

	public int getTimeOnStartingToBroadcast() {
		return timeOnStartingToBroadcast;
	}

	public void setTimeOnStartingToBroadcast(int timeOnStartingToBroadcast) {
		this.timeOnStartingToBroadcast = timeOnStartingToBroadcast;
	}

	public int getTimeOnEquipingToBroadcast() {
		return timeOnEquipingToBroadcast;
	}

	public void setTimeOnEquipingToBroadcast(int timeOnEquipingToBroadcast) {
		this.timeOnEquipingToBroadcast = timeOnEquipingToBroadcast;
	}

	public String getMessagePrefix() {
		return messagePrefix;
	}

	public void setMessagePrefix(String messagePrefix) {
		this.messagePrefix = messagePrefix;
	}

	public List<MinigameLobby> getLobbies() {
		return lobbies;
	}

	public void setLobbies(List<MinigameLobby> lobbies) {
		this.lobbies = lobbies;
	}

	public int getMaxPlayersPerLobby() {
		return maxPlayersPerLobby;
	}

	public void setMaxPlayersPerLobby(int maxPlayersPerLobby) {
		this.maxPlayersPerLobby = maxPlayersPerLobby;
	}

}

File: src/main/java/net/eduard/api/server/minigame/Minigame.kt
package net.eduard.api.server.minigame

import org.bukkit.Location
import org.bukkit.entity.Player

import net.eduard.api.lib.manager.TimeManager
import net.eduard.api.lib.kotlin.reloadListFromFolder
import net.eduard.api.lib.kotlin.saveListInFolder
import net.eduard.api.lib.modules.BukkitBungeeAPI
import net.eduard.api.lib.modules.FakePlayer
import net.eduard.api.lib.modules.Mine
import net.eduard.api.lib.score.DisplayBoard
import net.eduard.api.lib.storage.annotations.StorageIndex
import java.io.File

/**
 * Representa um Minigame ou Evento <br></br>
 * Nomes anteriores<br>
 * MinigameSetup 1.0
 *
 * @version 2.0
 * @since EduardAPI 2.0
 * @author Eduard
 */
@Suppress("unused")

open class Minigame(
    @StorageIndex
    var name: String,
    var messagePrefix: String
) : TimeManager() {

    constructor() : this( "Minigame")
    constructor(name: String) : this(name,"§8[§b$name§8]§f")

    var isEnabled = true
    var isBungeecord = false
    var schematicBased = true

    var worldBased
        get() = !schematicBased
        set(value) {
            schematicBased = !value
        }
    var schematicBasedRestartWorld = false
    var isLobby = false
    var bungeeLobby = "Lobby"
    var scoreboardLobby = DisplayBoard("Minigame lobby")
    var maxPlayersPerLobby = 20

    @Transient
    var setting: MinigameMap? = null
    var lobby: Location? = null
    var mapsWorld = MinigameWorld("worlds/$name/map_creation")

    @Transient
    var players: MutableMap<FakePlayer, MinigamePlayer> = mutableMapOf()


    @Transient
    var lobbies = mutableListOf<MinigameLobby>()

    @Transient
    var maps = mutableListOf<MinigameMap>()

    @Transient
    var rooms = mutableListOf<MinigameRoom>()

    @Transient
    var modes = mutableListOf<MinigameMode>()

    /**
     * Pega a primera sala existente do Minigame
     *
     * @return Sala
     */
    //		getRooms().get(0);
    open val game: MinigameRoom?
        get() = rooms.firstOrNull()

    open val mainLobby: MinigameLobby
        get() = if (lobbies.isNotEmpty()) lobbies[0] else createLobby(1)
    open val mode: MinigameMode?
        get() = modes.firstOrNull()

    /**
     * Pega o mapa referente a sala principal do Minigame
     *
     * @return
     */
    val map: MinigameMap?
        get() = getMap(name)

    fun getMode(name: String) = modes.firstOrNull { it.name.equals(name, true) }


    val isSetting: Boolean
        get() = setting != null

    /**
     * Conecta todos jogadores no servidor Lobby
     */
    open fun connectAllPlayersToLobby() {
        for (player in players.values) {
            BukkitBungeeAPI.connectToServer(player.player, bungeeLobby)
        }

    }

    /**
     * Registra o Modo de Jogo
     */
    fun register(mode: MinigameMode) {
        modes.add(mode)
        mode.minigame = this

    }

    /**
     * Teleporta todos os jogadores para o Local do Lobby
     */
    fun teleportAllPlayersToLobby() {
        lobby ?: players.values.forEach { it.player.teleport(lobby) }
    }


    /**
     * Cria um Mapa
     *
     * @param nome Nome
     * @return Mapa Novo
     */
    fun createMap(nome: String): MinigameMap {
        val map = MinigameMap(nome)
        map.minigame = this
        maps.add(map)
        return map
    }


    /**
     * Pega o mapa existente pelo seu nome
     *
     * @param name Nome
     * @return Mapa
     */
    open fun getMap(name: String): MinigameMap? {
        return maps.firstOrNull { it.name.equals(name, true) }
    }

    /**
     * Remove o mapa da lista de mapas existentes
     *
     * @param map
     */
    fun removeMap(map: MinigameMap) {
        maps.remove(map)
    }

    /**
     * Verifica se existe este Map com este Nome
     *
     * @param name Nome
     * @return se encontrou o mapa
     */
    fun hasMap(name: String): Boolean {
        return getMap(name) != null
    }

    /**
     * Cria um sala unica e um mapa unico Também usando o Nome do Minigame
     *
     * @return Minigame criado com mapa já configurado
     */
    open fun uniqueGame(): MinigameRoom {
        return createRoom(createMap(name), mode!!)
    }


    /**
     * Manda mensagem para todos jogadores participando do minigame
     *
     * @param message Mensagem
     */
    fun broadcast(message: String) {
        for (minigamePlayer in players.values) {
            minigamePlayer.sendMessage(
                messagePrefix +
                        Mine.getReplacers(message, minigamePlayer.player)
            )

        }
    }

    /**
     * Pega a sala pelo seu ID
     *
     * @param id ID
     * @return Sala
     */
    open fun getRoom(id: Int) = rooms.firstOrNull { it.id == id }


    /**
     * Verifica se a sala com este ID existe
     *
     * @param id ID
     * @return Sala
     */
    open fun hasRoom(id: Int) = getRoom(id) != null


    /**
     * Cria uma Sala com este ID para o Mapa expecifico
     *
     * @param map Mapa
     * @param id  ID
     * @return Nova Sala
     */
    open fun createRoom(map: MinigameMap, mode: MinigameMode): MinigameRoom {
        val room = MinigameRoom()
        room.minigame = this
        room.id = rooms.size + 1
        room.mode = mode
        rooms.add(room)
        room.map = map
        this.isEnabled = true
        return room
    }


    /**
     * Pega a sala que o jogador esta jogando
     *
     * @param player Jogador
     * @return Sala do jogador
     */
    open fun getGame(player: Player): MinigameRoom? {
        return getPlayer(player).game
    }

    /**
     * Pega a sala com o nome do seu mapa igual a este
     *
     * @param name Nome
     * @return
     */
    open fun getGame(name: String): MinigameRoom? {
        for (room in rooms) {
            if (room.map.name.equals(name, ignoreCase = true)) {
                return room
            }
        }
        return null
    }

    /**
     * Cria um lobby com ID novo
     */
    open fun createLobby(id: Int): MinigameLobby {

        val lobby = MinigameLobby()
        lobby.id = id
        lobbies.add(lobby)
        return lobby

    }


    /**
     * Pega o MinigamePlayer referente ao jogador e se caso não exista cria um
     *
     * @param player Jogador
     * @return Instancia de MinigamePlayer (MP)
     */
    fun getPlayer(player: FakePlayer): MinigamePlayer {
        var member = players[player]
        if (member == null) {
            member = MinigamePlayer()
            member.fakePlayer = player
            players[player] = member
        }



        return member
    }

    /**
     * Alias para o getPlayer(FakePlayer())
     */
    fun getPlayer(player: Player): MinigamePlayer {
        return getPlayer(FakePlayer(player))
    }

    /**
     * Verifica se existe o lobby
     *
     * @return
     */
    fun hasLobby(): Boolean {
        return lobby != null
    }

    /**
     * Verifica se o jogador esta no modo Admin
     *
     * @param player Jogador
     * @return
     */
    fun isAdmin(player: Player): Boolean {
        return getPlayer(player).isState(MinigamePlayerState.ADMIN)

    }

    /**
     * Verifica se o Jogador esta no modo Normal (sem ser Admin ou Spectador)
     *
     * @param player Jogador
     * @return se sim ou não
     */
    fun isPlayer(player: Player): Boolean {
        return getPlayer(player).isState(MinigamePlayerState.NORMAL)

    }

    /**
     * Verifica se o jogador esta no Minigame
     *
     * @param player Jogador
     * @return
     */
    fun isPlaying(player: Player): Boolean {
        return getPlayer(player).isPlaying
    }

    /**
     * Verifica se o Jogador esta no modo Spectador
     *
     * @param player Jogador
     * @return
     */
    fun isSpectator(player: Player): Boolean {
        return getPlayer(player).isState(MinigamePlayerState.SPECTATOR)

    }

    /**
     * Verifica se o Estado da Sala principal é igual este estado
     *
     * @param state Estado
     * @return
     */
    fun isState(state: MinigameState): Boolean {
        return game?.isState(state) ?: false
    }

    /**
     * Entrar em uma Sala
     *
     * @param game   Sala
     * @param player Jogador
     */
    fun joinPlayer(game: MinigameRoom, player: Player) {
        val minigamePlayer = getPlayer(player)
        minigamePlayer.join(game)

    }

    /**
     * Entrar em um Time
     *
     * @param team   Time
     * @param player Jogador
     */
    fun joinPlayer(team: MinigameTeam, player: Player) {
        val minigamePlayer = getPlayer(player)
        minigamePlayer.join(team)
    }

    /**
     * Remover o jogador da sala e do time Atual dele
     *
     * @param player Jogador
     */
    fun leavePlayer(player: Player) {
        val minigamePlayer = getPlayer(player)
        if (minigamePlayer.isPlaying) {
            minigamePlayer.game?.leave(minigamePlayer)
        }

        if (minigamePlayer.hasTeam()) {
            minigamePlayer.team?.leave(minigamePlayer)
        }

    }

    /**
     * Remove o jogador da HashMap de jogadores [MinigamePlayer]
     *
     * @param player Jogador
     */
    open fun remove(player: Player) {
        players.remove(FakePlayer(player))
    }

    /**
     * Remove a Sala da lista de salas existentes
     *
     * @param game Sala
     */
    open fun removeGame(game: MinigameRoom) {
        this.rooms.remove(game)
    }

    /**
     * Remove a sala pelo seu ID
     *
     * @param id
     */
    open fun removeGame(id: Int) {
        val game = getRoom(id)
        this.rooms.remove(game)
    }


    /**
     * Metodo que é executado a cada segundo e executa o metodo de cada sala
     * ` listener(room)`
     */
    override fun run() {
        if (!isEnabled)
            return
        for (room in rooms) {
            if (!room.isEnabled)
                continue
            room.mode.event(room)
        }
    }

    open fun save() {
        saveLobbies()
        saveModes()
        saveMaps()
        saveRooms()

    }

    fun saveModes() {
        for (mode in modes) {
            mode.saveChests()
            mode.saveKits()
        }
        File(plugin.dataFolder, "modes/")
            .saveListInFolder(modes) { it.modeName }
    }


    fun saveRooms() {
        File(plugin.dataFolder, "rooms/")
            .saveListInFolder(rooms) { it.id.toString() }
    }
    open fun hasMinPlayerAmount(room : MinigameRoom) : Boolean{
        return room.getPlayers(MinigamePlayerState.NORMAL).size >= room.map.minPlayersAmount
    }

    fun saveLobbies() {
        File(plugin.dataFolder, "lobby/")
            .saveListInFolder(lobbies) {
                it.id.toString()
            }

    }


    fun saveMaps() {
        File(plugin.dataFolder, "maps/").saveListInFolder(maps) {
            name.toLowerCase()
                .replace(" ", "_")
        }

    }

    open fun reload() {
        reloadLobbies()
        reloadMaps()
        reloadModes()
        reloadRooms()
    }

    fun reloadModes() {
        modes.clear()
        modes.addAll(File(plugin.dataFolder, "modes/").reloadListFromFolder())
        modes.forEach {
            it.minigame = this
            it.reloadChests()
            it.reloadKits()
        }

    }

    fun reloadMaps() {
        maps.clear()
        maps.addAll(File(plugin.dataFolder, "maps/").reloadListFromFolder())
        maps.forEach { it.minigame = this }
    }

    fun reloadLobbies() {
        lobbies.clear()
        lobbies.addAll(File(plugin.dataFolder, "lobby/").reloadListFromFolder())
    }


    fun reloadRooms() {
        rooms.clear()
        rooms.addAll(File(plugin.dataFolder, "rooms/").reloadListFromFolder())
        rooms.forEach { it.minigame = this }
    }


}